transport/src/main/java/io/netty/channel/embedded/EmbeddedChannel.java

```diff
@@ -200,10 +204,21 @@ public EmbeddedChannel(Channel parent, ChannelId channelId, boolean register, bo
      */
     public EmbeddedChannel(ChannelId channelId, boolean hasDisconnect, final ChannelConfig config,
                            final ChannelHandler... handlers) {
-        super(null, channelId);
-        metadata = metadata(hasDisconnect);
-        this.config = ObjectUtil.checkNotNull(config, "config");
-        setup(true, handlers);
+        this(builder().channelId(channelId).hasDisconnect(hasDisconnect).config(config).handlers(handlers));
+    }
+
+    /**
+     * Create a new instance with the configuration from the given builder. This method is {@code protected} for use by
+     * subclasses; Otherwise, please use {@link Builder#build()}.
+     *
+     * @param builder The builder
+     */
+    protected EmbeddedChannel(Builder builder) {
```

Comment:
why not package-private @yawkat ?

Reply:
It is necessary to extend EmbeddedChannel sometimes (see https://github.com/netty/netty/pull/12546#issuecomment-1172856534 ), and subclasses should also have the option of using all the builder properties.